### ISSUE TITLE
New version with only CPU support

### ISSUE THREAD

**Issue by JeanMarcAINow** [aluo-x/tensorflow_windows#1]
My laptop has no GPU.
Would it possible to build another version with only CPU support?

**Comment by aluo-x** [comment]
Sure, I can probably create a build this Wednesday.

**Comment by aluo-x** [comment]
I've uploaded the CPU only version compiled with AVX support. I decided to omit AVX2 support since there have been reports of failed builds. 

Please test it out and let me know if there are any issues.

**Comment by JeanMarcAINow** [comment]
Thanks Andrew,
I have cloned my conda environment to test your 2 Tensorflow versions and compare them to the regular one.
Unfortunately it does not work for any of them and it is also crashing Python.

I have attached the Jupyter notebook and its related data, the error message and my technical configuration.

[Testing Tensorflow Version.zip](https://github.com/aluo-x/tensorflow_windows/files/1116968/Testing.Tensorflow.Version.zip)

**Comment by aluo-x** [comment]
I am going through your notebook, however I am not sure if the error is from tensorflow or the interfacing between keras&tensorflow. 

Could you try running [inception](https://github.com/soumith/convnet-benchmarks/blob/master/tensorflow/benchmark_googlenet.py), maybe using a small batch size of <20 to avoid memory issues, and let me know if it runs?

**Comment by JeanMarcAINow** [comment]
The regullar ones on PIP and Conda work. They were version 1.1 for Python 3.5 and 3.6.
I saw this morning a new version from Google 1.2.1 that I have installed on Python 3.6.
It gives the usual warning about compilation optimisations. I have attached the famous Martin Görner python sources for MNIST if you want to test.

root) D:\Data\Machine Learning\Python\MNIST\Archives\Tensorflow and deep learning - without a PhD by Martin Görner\tensorflow-mnist-tutorial-master>python mnist_4.2_batchnorm_convolutional.py
Tensorflow version 1.2.1
Extracting data\train-images-idx3-ubyte.gz
Extracting data\train-labels-idx1-ubyte.gz
Extracting data\t10k-images-idx3-ubyte.gz
Extracting data\t10k-labels-idx1-ubyte.gz
2017-07-02 07:01:34.967308: W c:\tf_jenkins\home\workspace\release-win\m\windows\py\36\tensorflow\core\platform\cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE instructions, but these are available on your machine and could speed up CPU computations.
.....

[Tensorflow and deep learning - without a PhD by Martin Görner.zip](https://github.com/aluo-x/tensorflow_windows/files/1117609/Tensorflow.and.deep.learning.-.without.a.PhD.by.Martin.Gorner.zip)

**Comment by JeanMarcAINow** [comment]
You have updated your CPU version 3 days ago that I did not have.

I have removed all tensorflow packages from my conda testing environment and installed your latest version.
I have attached the error message.

[error message.txt](https://github.com/aluo-x/tensorflow_windows/files/1117675/error.message.txt)

**Comment by JeanMarcAINow** [comment]
I did one more with updating all packages, but no change and still the same error.

**Comment by aluo-x** [comment]
I am investigating, can you try and run [Inception](https://github.com/soumith/convnet-benchmarks/blob/master/tensorflow/benchmark_googlenet.py) using a small batch size of maybe around 10 and see if it works? Trying to figure out if this is some kind Keras bug or tensorflow issue.

**Comment by JeanMarcAINow** [comment]
I have run the script benchmark_googlenet.py from your link but it fails on all Tensorflow versions (official and yours) with the following error:

InvalidArgumentError (see above for traceback): CPU BiasOp only supports NHWC.
         [[Node: conv1/BiasAdd = BiasAdd[T=DT_FLOAT, data_format="NCHW", _device="/job:localhost/replica:0/task:0/cpu:0"](conv1/Conv2D, conv1/biases/read)]]

All the scripts attached previously have been working fine on officials Tensorflow versions (1.1 and 1.2, and also on my 2 laptops, CPU and GPU versions).

I am new to the field and I am currently experimenting with the MNIST challenge.
I can run any script you need but in the scope of my limited skills.

**Comment by JeanMarcAINow** [comment]
Here are the results. No crash on any version.
Your version is hosted on Python35CPU at the end.

3 environments:
default: (C:\Users\jeanm\Miniconda3) is Python 3.6.1 - tensorflow Version: 1.2.1
(stable) is Python 3.5.3 - tensorflow Version: 1.1.0
(python35CPU) is Python 3.5.3 - tensorflow Version: 1.2.0 - Optimized build

(C:\Users\jeanm\Miniconda3) 

2017-07-03 07:45:02.957182: step 10, duration = 1.215
2017-07-03 07:45:15.028786: step 20, duration = 1.204
2017-07-03 07:45:27.054816: step 30, duration = 1.202
2017-07-03 07:45:39.083853: step 40, duration = 1.201
2017-07-03 07:45:51.110153: step 50, duration = 1.194
2017-07-03 07:46:03.153233: step 60, duration = 1.201
2017-07-03 07:46:15.162262: step 70, duration = 1.200
2017-07-03 07:46:27.189807: step 80, duration = 1.205
2017-07-03 07:46:39.204633: step 90, duration = 1.199
2017-07-03 07:46:50.021954: Forward across 100 steps, 1.192 +/- 0.120 sec / batch
2017-07-03 07:48:10.875563: step 10, duration = 3.756
2017-07-03 07:48:48.591930: step 20, duration = 3.699
2017-07-03 07:49:25.581707: step 30, duration = 3.694
2017-07-03 07:50:02.811990: step 40, duration = 3.707
2017-07-03 07:50:39.816540: step 50, duration = 3.712
2017-07-03 07:51:17.142095: step 60, duration = 3.765
2017-07-03 07:51:54.636539: step 70, duration = 3.752
2017-07-03 07:52:32.324904: step 80, duration = 3.735
2017-07-03 07:53:09.368709: step 90, duration = 3.717
2017-07-03 07:53:43.029759: Forward-backward across 100 steps, 3.698 +/- 0.374 sec / batch

(stable) 

2017-07-03 07:55:21.195724: step 10, duration = 1.208
2017-07-03 07:55:33.199198: step 20, duration = 1.203
2017-07-03 07:55:45.211149: step 30, duration = 1.206
2017-07-03 07:55:57.139849: step 40, duration = 1.196
2017-07-03 07:56:09.058530: step 50, duration = 1.195
2017-07-03 07:56:21.038378: step 60, duration = 1.233
2017-07-03 07:56:33.268705: step 70, duration = 1.198
2017-07-03 07:56:45.395695: step 80, duration = 1.224
2017-07-03 07:56:57.433807: step 90, duration = 1.200
2017-07-03 07:57:08.179881: Forward across 100 steps, 1.190 +/- 0.121 sec / batch
2017-07-03 07:58:29.107072: step 10, duration = 3.775
2017-07-03 07:59:07.259460: step 20, duration = 3.786
2017-07-03 07:59:45.551720: step 30, duration = 4.070
2017-07-03 08:00:25.975644: step 40, duration = 3.753
2017-07-03 08:01:03.653771: step 50, duration = 3.859
2017-07-03 08:01:41.247675: step 60, duration = 3.731
2017-07-03 08:02:18.831301: step 70, duration = 3.827
2017-07-03 08:02:57.826582: step 80, duration = 4.253
2017-07-03 08:03:36.462045: step 90, duration = 3.817
2017-07-03 08:04:11.373983: Forward-backward across 100 steps, 3.802 +/- 0.430 sec / batch

(python35CPU) 

2017-07-03 08:11:36.011652: step 10, duration = 0.766
2017-07-03 08:11:43.529286: step 20, duration = 0.754
2017-07-03 08:11:51.014714: step 30, duration = 0.750
2017-07-03 08:11:58.526821: step 40, duration = 0.753
2017-07-03 08:12:06.315589: step 50, duration = 0.754
2017-07-03 08:12:14.012291: step 60, duration = 0.749
2017-07-03 08:12:21.840233: step 70, duration = 0.753
2017-07-03 08:12:29.375085: step 80, duration = 0.750
2017-07-03 08:12:36.973605: step 90, duration = 0.764
2017-07-03 08:12:43.829868: Forward across 100 steps, 0.755 +/- 0.079 sec / batch
2017-07-03 08:13:40.301563: step 10, duration = 2.521
2017-07-03 08:14:06.217931: step 20, duration = 2.580
2017-07-03 08:14:33.229641: step 30, duration = 2.591
2017-07-03 08:14:58.784918: step 40, duration = 2.515
2017-07-03 08:15:23.759988: step 50, duration = 2.490
2017-07-03 08:15:48.699167: step 60, duration = 2.523
2017-07-03 08:16:13.644652: step 70, duration = 2.497
2017-07-03 08:16:38.522270: step 80, duration = 2.501
2017-07-03 08:17:03.487117: step 90, duration = 2.480
2017-07-03 08:17:25.993758: Forward-backward across 100 steps, 2.524 +/- 0.279 sec / batch

**Comment by aluo-x** [comment]
I have build a [pure tensorflow based model](https://gist.github.com/aluo-x/46ec50350a991c749308c1ce63c75bd3) of your first mnist code, I have not had time to really check the code, so there may be bugs. It downloads mnist data into C:\tmp.

I am certain the error is coming from a bug in batchnorm, because after removing that the code worked fine. Not sure if this is a tensorflow bug or a bug in compilation, but I would recommend that you stop using this code unless the speed improvement is really worth it.

**Comment by aluo-x** [comment]
This is so bizarre, the failure has been very non deterministic.  
For the same code, I have encountered a few different errors.
c:\optimae\tensorflow-1.2.1\tensorflow\core\framework\tensor_shape.cc:170] Check failed: ndims_byte() < MaxDimensions() (unsigned char value 255 vs. 255)Too many dimensions in tensor

**Comment by aluo-x** [comment]
I have a build that is CPU only and compiled with AVX support. I "believe" this fixes the issues that you were experiencing. There have been two changes, for this build I am using the default python35.lib instead of the intel optimized version, and I have switched to VS2017 instead of VS2015. If I were to guess, I would say it was probably the intel python35.lib that was the culprit.

**Comment by FrozenKiwi** [comment]
Hi,

Thanks a tonne for building this!  However - after installation I'm not seeing any performance increase training a simple 1-50-1 NN?

I expected to see nearly 4x improvements, so I'm a bit confused why there is no change.  If anything, it feels a touch slower...

Have you verified it's giving you the speed increase you expected?

**Comment by FrozenKiwi** [comment]
And the comment I -actually- meant to ask :-) - do you plan to do windows build for CUDA again?

Thanks again for what you have done here!

**Comment by aluo-x** [comment]
Here are the benchmarks, using inception, a batch size of 10, and NHWC ordering. File used is [found here](https://raw.githubusercontent.com/soumith/convnet-benchmarks/master/tensorflow/benchmark_googlenet.py).
This was tested on a laptop with an i7-7700HQ clocked at 3.4 GHz. The system was allowed sufficient time to cool down between the tests.

Generic version from Google:
2017-07-14 01:01:10.115939: step 10, duration = 0.500
2017-07-14 01:01:15.034729: step 20, duration = 0.500
2017-07-14 01:01:19.925867: step 30, duration = 0.500
2017-07-14 01:01:24.848257: step 40, duration = 0.484
2017-07-14 01:01:29.755020: step 50, duration = 0.484
2017-07-14 01:01:34.650724: step 60, duration = 0.500
2017-07-14 01:01:39.557487: step 70, duration = 0.484
2017-07-14 01:01:44.459934: step 80, duration = 0.484
2017-07-14 01:01:49.382323: step 90, duration = 0.484
2017-07-14 01:01:53.804661: Forward across 100 steps, 0.486 +/- 0.049 sec / batch
2017-07-14 01:02:27.975038: step 10, duration = 1.563
2017-07-14 01:02:43.617960: step 20, duration = 1.564
2017-07-14 01:02:59.291477: step 30, duration = 1.547
2017-07-14 01:03:14.931660: step 40, duration = 1.547
2017-07-14 01:03:30.495789: step 50, duration = 1.563
2017-07-14 01:03:46.127817: step 60, duration = 1.578
2017-07-14 01:04:01.832549: step 70, duration = 1.594
2017-07-14 01:04:17.537381: step 80, duration = 1.563
2017-07-14 01:04:33.179642: step 90, duration = 1.547
2017-07-14 01:04:47.270391: Forward-backward across 100 steps, 1.549 +/- 0.156 sec / batch

My build w/ AVX enabled:
2017-07-14 00:47:43.761157: step 10, duration = 0.344
2017-07-14 00:47:47.245898: step 20, duration = 0.344
2017-07-14 00:47:50.657750: step 30, duration = 0.328
2017-07-14 00:47:54.033105: step 40, duration = 0.328
2017-07-14 00:47:57.427897: step 50, duration = 0.328
2017-07-14 00:48:00.850129: step 60, duration = 0.344
2017-07-14 00:48:04.244983: step 70, duration = 0.344
2017-07-14 00:48:07.651590: step 80, duration = 0.344
2017-07-14 00:48:11.058196: step 90, duration = 0.344
2017-07-14 00:48:14.089763: Forward across 100 steps, 0.337 +/- 0.035 sec / batch
2017-07-14 00:48:40.458270: step 10, duration = 1.203
2017-07-14 00:48:52.495525: step 20, duration = 1.219
2017-07-14 00:49:04.441662: step 30, duration = 1.203
2017-07-14 00:49:16.381204: step 40, duration = 1.157
2017-07-14 00:49:28.332074: step 50, duration = 1.219
2017-07-14 00:49:40.270824: step 60, duration = 1.188
2017-07-14 00:49:52.180064: step 70, duration = 1.188
2017-07-14 00:50:04.174511: step 80, duration = 1.203
2017-07-14 00:50:16.097634: step 90, duration = 1.188
2017-07-14 00:50:26.899524: Forward-backward across 100 steps, 1.184 +/- 0.120 sec / batch

This works out to be (1/0.337)/(1/0.486) -1 = 44% faster for forward, (1/1.184)/(1/1.549)-1 = 30% faster for backwards prop. I am actually more concerned to see if my build is functionally the same as the generic build, since the previous VS 2015 + Intel Python version was corrupted.

CUDA builds are currently failing, kind of a miracle that the CPU build actually worked at all. Will keep experimenting to see if I can get a GPU + AVX build from VS 2017.

**Comment by JeanMarcAINow** [comment]
Thanks aluo-x

I have quickly tested your new version and I haven't seen any crash anymore.
I will let you know if I notice any issue.
I can confirm that this custom built version is faster on CPU than the standard one by one third.
My previous results above are for an i3-6157 clocked at 2.4 GHz.

**Comment by FrozenKiwi** [comment]
Thanks for the answer aluo-x.  It's odd 40% should be visible in my set.  Perhaps my problem has a different bottleneck.

Dell XPS 15 w/ i7 4712HQ @2.30, and a bad tendency to overheat - so not a beast by any stretch.

On the flip side - it seems to be working fine for me too  - I'll be sticking with this build for the time being.

**Comment by aluo-x** [comment]
GPU build with AVX showed no noticeable speed increased on my laptop and has broken batchnorm. However they now have CUDA SM30, SM52 and SM61. Closing this issue for now. Feel free to reopen if necessary.

**Comment by aluo-x** [comment]
Heads up, Tensorflow 1.4 with AVX2 (Intel Haswell and later, AMD Excavator & Zen) has been posted.

**Comment by aluo-x** [comment]
a CUDA 8 + cudnn 7 TF 1.4 with AVX support has been uploaded. Let me know if it works
